onnx/test/test_external_data.py

```diff
@@ -342,49 +342,186 @@ def test_save_model_does_convert_and_saves_the_model(self):  # type: () -> None
+        self.assertTrue(np.allclose(to_array(w_small_tensor), data_w_small))
+
+        w_large_tensor = model.graph.initializer[1]
+        self.assertTrue(w_large_tensor.HasField("data_location"))
+        self.assertTrue(np.allclose(to_array(w_large_tensor, self.temp_dir), data_w_large))
+
+    def test_to_array_with_external_data(self):  # type: () -> None
+        x = data_w = np.ones((3, 3), np.float32)
+        w = helper.make_tensor(name='w', data_type=TensorProto.FLOAT, dims=data_w.shape, vals=data_w.flatten().astype(np.float32).tobytes(), raw=True)
+        V = helper.make_tensor_value_info('V', TensorProto.FLOAT, [3, 3])
+        Z = helper.make_tensor_value_info('Z', TensorProto.FLOAT, [3, 3])
+
+        X = helper.make_node(
+            'Constant',
+            inputs=[],
+            outputs=['X'],
+            value=onnx.helper.make_tensor(
+                name='const_x',
+                data_type=onnx.TensorProto.FLOAT,
+                dims=x.shape,
+                vals=x.flatten().astype(float),
+            ),
+        )
+        node_cf = helper.make_node(
+            'Gemm',
+            ['w', 'X'],
+            ['Z'],
+            name='gemm'
+        )
+        graph_def = helper.make_graph(
+            [X, node_cf],
+            'test-model',
+            [V],
+            [Z],
+            initializer=[w],
+        )
+        model_def = helper.make_model(graph_def, producer_name='onnx-example')
+
+        path = os.path.join(self.temp_dir, 'temp.onnx')
+        onnx.save_model(model_def, path, save_as_external_data=True, all_tensors_to_one_file=False, size_threshold=0)
+        # raw_data of external tensor is not loaded
+        model = onnx.load(path, load_external_data=False)
+        # Specify self.temp_dir to load external tensor
+        tensor_data = to_array(model.graph.initializer[0], self.temp_dir)
+        self.assertTrue(np.allclose(tensor_data, data_w))
+
+    def test_save_model_with_existing_raw_data_should_override(self):  # type: () -> None
+        model_file_path = self.get_temp_model_filename()
+        original_raw_data = self.model.graph.initializer[0].raw_data
+        onnx.save_model(self.model, model_file_path, save_as_external_data=True, size_threshold=0)
+        self.assertTrue(Path.isfile(model_file_path))
+
+        model = onnx.load_model(model_file_path, load_external_data=False)
+        initializer_tensor = model.graph.initializer[0]
+        initializer_tensor.raw_data = b'dummpy_raw_data'
+        # If raw_data and external tensor exist at the same time, override existing raw_data
+        load_external_data_for_tensor(initializer_tensor, self.temp_dir)
+        self.assertEqual(initializer_tensor.raw_data, original_raw_data)
+
+    def test_reshape_inference_with_external_data_fail(self):  # type: () -> None
+        reshape_shape = (2, 12)
+        X = helper.make_tensor_value_info('X', TensorProto.FLOAT, [2, 3, 4])
```

Comment:
see if you can move the model creation code to a utility function and call it from all these tests... atleast a couple of tests can share the same model

Reply:
I have made these 3 tests (test_reshape_inference_with_external_data_fail, test_to_array_with_external_data, test_save_model_with_external_data_multiple_times) using the same model, which simplifies the code. Thanks for pointing out.